diff -Naur --exclude=CVS --exclude='*.o' --exclude='*.a' --exclude='*.so' --exclude='*.elf' --exclude=System.map --exclude=Makefile.d --exclude='*log' --exclude='*log2' --exclude='*~' --exclude='.*~' --exclude='.#*' --exclude='*.bak' --exclude='*.orig' --exclude='*.rej' --exclude='core.[0-9]*' --exclude=.depend --exclude='.*.o.flags' --exclude='*.gz' --exclude=.depend --exclude='.*.o.flags' --exclude='*.gz' --exclude=vmlinux --exclude=vmlinux.bin --exclude=yamon-02.06-SIGMADESIGNS-01_el.bin linux-2.6.22.16.ref/net/core/dev.c linux-2.6.22.16/net/core/dev.c
--- linux-2.6.22.16.ref/net/core/dev.c	2008-01-14 17:17:32.000000000 -0800
+++ linux-2.6.22.16/net/core/dev.c	2008-03-04 19:12:29.000000000 -0800
@@ -453,6 +453,9 @@
 	return 0;
 }
 
+#ifdef CONFIG_SD_IPFILTER
+void *(*sd_ipfilter)(unsigned char *, int) = NULL;
+#endif
 
 /**
  *	netdev_boot_base	- get address from boot time settings
@@ -1640,6 +1643,13 @@
 	struct softnet_data *queue;
 	unsigned long flags;
 
+#ifdef CONFIG_SD_IPFILTER
+	if ((sd_ipfilter != NULL) && (((*sd_ipfilter)(skb->data, skb->len)) == NULL)) {
+		kfree_skb(skb);
+		return NET_RX_SUCCESS;
+	}
+#endif
+
 	/* if netpoll wants it, pretend we never saw it */
 	if (netpoll_rx(skb))
 		return NET_RX_DROP;
@@ -1838,6 +1848,13 @@
 	int ret = NET_RX_DROP;
 	__be16 type;
 
+#ifdef CONFIG_SD_IPFILTER
+	if ((sd_ipfilter != NULL) && (((*sd_ipfilter)(skb->data, skb->len)) == NULL)) {
+		kfree_skb(skb);
+		return NET_RX_SUCCESS;
+	}
+#endif
+
 	/* if we've gotten here through NAPI, check netpoll */
 	if (skb->dev->poll && netpoll_rx(skb))
 		return NET_RX_DROP;
@@ -3785,4 +3802,8 @@
 EXPORT_SYMBOL(dev_load);
 #endif
 
+#ifdef CONFIG_SD_IPFILTER
+EXPORT_SYMBOL(sd_ipfilter);
+#endif
+
 EXPORT_PER_CPU_SYMBOL(softnet_data);
diff -Naur --exclude=CVS --exclude='*.o' --exclude='*.a' --exclude='*.so' --exclude='*.elf' --exclude=System.map --exclude=Makefile.d --exclude='*log' --exclude='*log2' --exclude='*~' --exclude='.*~' --exclude='.#*' --exclude='*.bak' --exclude='*.orig' --exclude='*.rej' --exclude='core.[0-9]*' --exclude=.depend --exclude='.*.o.flags' --exclude='*.gz' --exclude=.depend --exclude='.*.o.flags' --exclude='*.gz' --exclude=vmlinux --exclude=vmlinux.bin --exclude=yamon-02.06-SIGMADESIGNS-01_el.bin linux-2.6.22.16.ref/README.1010.ipfilter.patch linux-2.6.22.16/README.1010.ipfilter.patch
--- linux-2.6.22.16.ref/README.1010.ipfilter.patch	1969-12-31 16:00:00.000000000 -0800
+++ linux-2.6.22.16/README.1010.ipfilter.patch	2008-03-05 15:00:24.000000000 -0800
@@ -0,0 +1,19 @@
+Feature:
+--------
+Hooks for Sigma's IP filter.
+
+Prerequisite patch numbers:
+---------------------------
+0000
+1000
+
+Primary author:
+---------------
+YH Lin
+
+Related to which chip version SMP86xx xx=?
+-----------------------------------------
+all
+
+(linux patches) which CONFIG_... are provided:
+----------------------------------------------
