diff -ur LLTD-PortingKit/Sample_Code/src/mapping.c LLTD-PortingKit.modify/Sample_Code/src/mapping.c
--- LLTD-PortingKit/Sample_Code/src/mapping.c	2009-03-31 19:25:11.000000000 +0800
+++ LLTD-PortingKit.modify/Sample_Code/src/mapping.c	2009-11-06 20:18:12.000000000 +0800
@@ -81,6 +81,7 @@
 {
     struct timeval now;
 
+    pktlen -= 4;
     g_ctc_packets++;
     if (g_ctc_packets > TOPO_CTC_PACKETS_MAX)
 	g_ctc_packets = TOPO_CTC_PACKETS_MAX;
