--- /home/steven/job/avls700l/src/libs/3_6_1_rc2/linux_kernel_2.6.22.19-12/drivers/char/Makefile	2009-02-27 12:23:03.000000000 +0800
+++ drivers/char/Makefile	2009-04-02 20:25:55.000000000 +0800
@@ -97,6 +97,7 @@
 obj-$(CONFIG_GPIO_TB0219)	+= tb0219.o
 obj-$(CONFIG_TELCLOCK)		+= tlclk.o
 obj-$(CONFIG_TANGOX_IR)		+= irkernel.o
+obj-$(CONFIG_8AV500LE_GPIO)	+= 8av500le_gpio.o
 obj-$(CONFIG_TANGOX_FIP) 	+= fipkernel.o
 obj-$(CONFIG_TANGOX_SCARD) 	+= sckernel.o
 obj-$(CONFIG_TANGOX_UIR) 	+= uirkernel.o

--- drivers/char/Kconfig.ori	2009-05-07 18:58:45.000000000 +0800
+++ drivers/char/Kconfig	2009-05-07 19:05:45.000000000 +0800
@@ -459,6 +459,12 @@
 
 source "drivers/serial/Kconfig"
 
+config 8AV500LE_GPIO
+       tristate "Control GPIO of 8AV500LE board"
+       default n
+       help
+         Control GPIO of 8AV500LE board"
+
 config TANGOX_IR
        tristate "SMP86xx IR remote support" 
        depends on TANGOX
