--- ./drivers/net/Kconfig.org	2009-10-30 01:30:09.000000000 +0800
+++ ./drivers/net/Kconfig	2009-10-30 10:26:36.000000000 +0800
@@ -1924,6 +1924,12 @@
 	  This led chip brand is C+ and the model is IP101A. The option use
 	  MII set phy some register bit to turn on or off the ethernet LED.
 
+config TANGOX_ENET_FIX_UP_PACKAGE_LENGTH
+	bool "Decrease frame length by 4 which is CRC length"
+	depends on TANGOX_ENET
+	help
+	  This received package length includes 4 byte CRC. Problem happens if you read package in promiscuous mode.
+
 config ACENIC
 	tristate "Alteon AceNIC/3Com 3C985/NetGear GA620 Gigabit support"
 	depends on PCI
--- ./drivers/net/tangox_enet.c.org	2009-10-30 02:18:56.000000000 +0800
+++ ./drivers/net/tangox_enet.c	2009-10-30 10:35:58.000000000 +0800
@@ -282,6 +282,11 @@
 
 			} else {
 
+#ifdef CONFIG_TANGOX_ENET_FIX_UP_PACKAGE_LENGTH
+				/* chlee: decrese 4B CRC of the frame. */
+				if(len >= 4)
+					len -= 4;
+#endif
 				/* ok, seems  valid, adjust skb  proto and len
 				 * and give it to kernel */
 				skb->dev = dev;
